### PR TITLE
Fix getting CPU topology

### PR DESCRIPTION
**Pull request by katarzyna-z**
Related to #2468 and  kubernetes/kubernetes#89859

### CODE REVIEW

File: machine/topology_test.go [hunk 1]
```diff
@@ -206,6 +206,82 @@ func TestTopologyEmptySysFs(t *testing.T) {
...
+
+	_, _, err := GetTopology(sysFs)
+	assert.Nil(t, err)
+	// assert.Equal(t, 1, len(topology))
+	// assert.Equal(t, 2, numCores)
+
+	// topologyJSON, err := json.Marshal(topology)
+	// assert.Nil(t, err)
+
+	// expectedTopology := `[
```
Comment (tedyu) [inline machine/topology_test.go:246]: You can drop commented out code

File: utils/sysinfo/sysinfo.go [hunk 2]
```diff
@@ -233,6 +241,61 @@ func GetNodesInfo(sysFs sysfs.SysFs) ([]info.Node, int, error) {
...
+		}
+		node.Cores = cores
+
+		err = addCacheInfo(sysFs, &node)
+		if err != nil {
+			return nil, 0, err
+		}
+		nodes = append(nodes, node)
+	}
+	return nodes, len(cpusPaths), nil
```
Comment (tedyu) [inline utils/sysinfo/sysinfo.go:277]: please add comment above line 244 explaining the meaning of return values, or name the return variables

File: utils/sysinfo/sysinfo.go [hunk 3]
```diff
@@ -233,6 +241,61 @@ func GetNodesInfo(sysFs sysfs.SysFs) ([]info.Node, int, error) {
...
+}
+
+func getCpusByPhysicalPackageID(sysFs sysfs.SysFs, cpusPaths []string) (map[int][]string, error) {
+	cpuPathBySocket := make(map[int][]string, 0)
```
Comment (tedyu) [inline utils/sysinfo/sysinfo.go:281]: cpuPathBySocket -> cpuPathsBySocket

**Comment by dashpole** [comment]
```
I0406 15:49:23.291] >> checking go formatting
I0406 15:49:24.399] The following files are not properly formatted:
I0406 15:49:24.400] utils/sysinfo/sysinfo.go utils/sysinfo/sysinfo.go
```

**Comment by dims** [comment]
@katarzyna-z Do we fetch the huge pages stuff? https://github.com/google/cadvisor/pull/2468/files#diff-0b1a389e68641fb87106247aed23f143L152-L187

**Comment by dims** [comment]
@dashpole yep, `TestMachineInformationIsReturned` works fine in the environment where i saw failure earlier.

LGTM (we can iterate!)

File: utils/sysinfo/sysinfo.go [hunk 4]
```diff
@@ -233,6 +241,62 @@ func GetNodesInfo(sysFs sysfs.SysFs) ([]info.Node, int, error) {
...
+	nodes := []info.Node{}
+
+	cpusPaths, err := sysFs.GetCPUsPaths(cpusPath)
+	if err != nil {
+		return nil, 0, err
+	}
+	cpusCount := len(cpusPaths)
+
+	if cpusCount == 0 {
+		err = fmt.Errorf("Any CPU is not available, cpusPath: %s", cpusPath)
```
Comment (tedyu) [inline utils/sysinfo/sysinfo.go:254]: Any CPU -> no CPU

File: utils/sysinfo/sysinfo.go [hunk 5]
```diff
@@ -233,6 +241,62 @@ func GetNodesInfo(sysFs sysfs.SysFs) ([]info.Node, int, error) {
...
+		nodes = append(nodes, node)
+	}
+	return nodes, cpusCount, nil
+}
+
+func getCpusByPhysicalPackageID(sysFs sysfs.SysFs, cpusPaths []string) (map[int][]string, error) {
+	cpuPathsByPhysicalPackageID := make(map[int][]string, 0)
+	for _, cpuPath := range cpusPaths {
+
+		rawPhysicalPackageID, _ := sysFs.GetCPUPhysicalPackageID(cpuPath)
```
Comment (tedyu) [inline utils/sysinfo/sysinfo.go:285]: the error return should be checked

Comment (tedyu) [reply utils/sysinfo/sysinfo.go:285]: See #2474

**Comment by katarzyna-z** [comment]
> @katarzyna-z Do we fetch the huge pages stuff? https://github.com/google/cadvisor/pull/2468/files#diff-0b1a389e68641fb87106247aed23f143L152-L187

@dims hugepages are still fetch when nodes topology is available, here: https://github.com/google/cadvisor/blob/069f7e7eb63d86e5d8d62f60b454affe30cdfa9c/utils/sysinfo/sysinfo.go#L234

**Comment by katarzyna-z** [comment]
I tested this pull request on following system without CONFIG_NUMA=y:
```
uname -a
Linux localhost.localdomain 3.10.0-1062.18.1.el7.config_numa_n.x86_64 #1 SMP Mon Apr 6 03:21:27 CEST 2020 x86_64 x86_64 x86_64 GNU/Linux
```
and I compared responses on machine API requests in previous version of cAdvisor (commit `9921cb3`) before changes introduced in `c5a9232` with version from this pull request.

Responses on `/api/v2.0/machine`:
- [commit 9921cb3](https://gist.githubusercontent.com/katarzyna-z/81a817d6c4485ea90f4dd17438e8d8b1/raw/5aaea575c55b67313be4f58e6b323f5a5faca886/commit_9921cb3.json)
- [this pull request](https://gist.githubusercontent.com/katarzyna-z/a939a061a56deefbf16e79871e240978/raw/da2fbd9b6abcfb436cb96b5e165aca3d2359de53/pr_with_fixing_cpu_topology.json)

There are only differences related to newly added elements.

**Comment by dims** [comment]
@katarzyna-z Thanks for the diffs! So can we say that there is some behavior change that we need to document? namely we now need `CONFIG_NUMA=y` for hugepages to work. (this was not the case earlier)

**Comment by dims** [comment]
@katarzyna-z also, please consider filing a follow up PR if you think Ted's comments above are helpful to fix.

**Comment by katarzyna-z** [comment]
> @katarzyna-z Thanks for the diffs! So can we say that there is some behavior change that we need to document? namely we now need `CONFIG_NUMA=y` for hugepages to work. (this was not the case earlier)

Information about hugepages is still read from "/sys/kernel/mm/hugepages/", see [this](https://github.com/google/cadvisor/blob/069f7e7eb63d86e5d8d62f60b454affe30cdfa9c/machine/info.go#L86)
and this information is available per machine and available at this place on API:
```
{
  "num_cores": 72,
  "num_physical_cores": 18,
  "num_sockets": 2,
  "cpu_frequency_khz": 3900000,
  "memory_capacity": 422127910912,
  "memory_by_type": {
    "Non-volatile-RAM": {
      "capacity": 2168421613568,
      "dimm_count": 8
    },
    "Unbuffered-DDR4": {
      "capacity": 412316860416,
      "dimm_count": 12
    }
  },
  "nvm": {
    "memory_mode_capacity": 0,
    "app direct_mode_capacity": 0,
    "avg_power_budget": 0
  },
  "hugepages": [
    {
      "page_size": 1048576,
      "num_pages": 16
    },
    {
      "page_size": 2048,
      "num_pages": 2048
    }
  ],
...
```
When nodes are available information about hugepages is also read per node form "/sys/devices/system/node/node0/hugepages" and available on API in topology section:
```
"topology": [
    {
      "node_id": 0,
      "memory": 214411317248,
      "hugepages": [
        {
          "page_size": 1048576,
          "num_pages": 8
        },
        {
          "page_size": 2048,
          "num_pages": 1024
        }
      ],
      "cores": [
     ....
{
      "node_id": 1,
      "memory": 214748364800,
      "hugepages": [
        {
          "page_size": 1048576,
          "num_pages": 8
        },
        {
          "page_size": 2048,
          "num_pages": 1024
        }
      ],
      "cores": [
...
```

What I see from code analysis and tests on machines the behavior was not changed but please correct me if I am wrong.

**Comment by dims** [comment]
@katarzyna-z thanks for confirming. It's all good. ( peeking at k/k https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cadvisor/util.go#L49-L54 )

**Pull request merged**